### PR TITLE
Mongo source

### PR DESCRIPTION
**Pull request by MaerF0x0**
adds a new source-- mongo based of the mgo.Iter iterators

https://clever.atlassian.net/browse/DATA-714

### CODE REVIEW

File: sources/mongo/mongo.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   // Connect to mongo, get your collection then do the following:
+	 iter := collection.Find(nil).Limit(100).Iter()
+	 table := mongo.New(iter)
+*/
+
+import (
+	"gopkg.in/Clever/optimus.v3"
+)
+
+// Iter simulates the mgo.Iter interface so we can remain independent
```
Comment (azylman) [inline sources/mongo/mongo.go]: Maybe give the fully qualified import path of mgo? e.g. "gopkg.in/mgo.v2" in case someone doesn't know what it is (unlikely if they're using mongo and go, but still seems worthwhile)

File: sources/mongo/mongo.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// New returns a new Table that iterates over all the results of a mongo query.
+func New(iter Iter) optimus.Table {
+	s := &mongoSource{rows: make(chan optimus.Row)}
+	go s.start(iter)
+	return s
+}
+
+// mongoSource type matches the
```
Comment (azylman) [inline sources/mongo/mongo.go]: incomplete comment

File: sources/mongo/mongo.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		r := optimus.Row{}
+		if !iter.Next(&r) {
+			break
+		}
+		s.rows <- r
+	}
+	s.err = iter.Err()
+}
+
+// Rows returns the read side of the channel of optimus Rows from this mongo source
```
Comment (azylman) [inline sources/mongo/mongo.go]: You don't need to document these since they're private, but if you want to, a common pattern in Go (to avoid duplicating documentation) is to say something like "Rows implements the Table interface", "Err implements the Table interface", etc., basically telling people that the description of that method is in the interface documentation.

I usually don't say anything about these private methods since it's pretty straightforward.

File: sources/mongo/mongo_test.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	data := testData[0]
+	// Build the optimusTable
+	sourceTable := New(data.GivenIter)
+
+	// Assert the result rows and errors occur
+	assert.Equal(t, data.ExpectedRows, tests.GetRows(sourceTable))
+	assert.Equal(t, data.ExpectedErr, sourceTable.Err())
+}
+
+func TestErrorRows(t *testing.T) {
```
Comment (azylman) [inline sources/mongo/mongo_test.go]: This test case does the same thing as `TestSuccessRows`, but operates on the second element in your list of tests. Usually when you use a table-based test, you have one `Test` method that iterates over the tests and does the same thing for each. e.g.:

```go
func TestMongoSource(t *testing.T) {
  for i, data := range testData {
    sourceTable := New(data.GivenIter)
    assert.Equal ...
    assert.Equal ..
  }
}
```

I think the csv package has a pretty good example of table-based tests: http://golang.org/src/encoding/csv/reader_test.go

File: transforms/transforms.go [hunk 5]
```diff
@@ -97,6 +97,22 @@ var JoinType = joinStruct{Left: joinType{0}, Inner: joinType{1}}
 
 // Join returns a TransformFunc that joins Rows with another table using the specified join type.
 func Join(rightTable optimus.Table, leftHeader string, rightHeader string, join joinType) optimus.TransformFunc {
+	return JoinCustom(rightTable, leftHeader, rightHeader, join, mergeRows)
```
Comment (azylman) [inline transforms/transforms.go]: This shouldn't be in this PR. This was a mistake from my original commit. Can you revert this file to the same as master?

File: sources/mongo/mongo.go (hunk 1 above)
Comment (MaerF0x0) [inline sources/mongo/mongo.go]: 036517c

File: sources/mongo/mongo.go (hunk 2 above)
Comment (MaerF0x0) [inline sources/mongo/mongo.go]: 036517c

File: sources/mongo/mongo.go (hunk 3 above)
Comment (MaerF0x0) [inline sources/mongo/mongo.go]: 036517c

File: sources/mongo/mongo_test.go (hunk 4 above)
Comment (MaerF0x0) [inline sources/mongo/mongo_test.go]: i wanted to mark it as parallel, the `make` is a little slow already.

File: transforms/transforms.go (hunk 5 above)
Comment (MaerF0x0) [inline transforms/transforms.go]: ea0cd9d

File: sources/mongo/mongo_test.go (hunk 4 above)
Comment (azylman) [inline sources/mongo/mongo_test.go]: If we want to make it faster, we should change `make` to execute the tests in parallel, or change this to run each row in the table in its own goroutine, instead of repeating code.

Comment (azylman) [inline sources/mongo/mongo_test.go]: If you want to do it by changing `make` to execute in parallel, we should do it in a separate PR. If you want to do it by changing it to use its own goroutine, it's in scope for this

Comment (azylman) [inline sources/mongo/mongo_test.go]: (personally, I'd advocate for doing `make` in parallel in a separate PR, but that also doesn't have to be part of the scope of your project - we can file an issue and come back to it)

Comment (MaerF0x0) [inline sources/mongo/mongo_test.go]: Dried and filed: https://clever.atlassian.net/browse/DATA-789

2b700bb

**Comment by azylman** [comment]
lgtm
